umpf

```diff
@@ -1855,6 +1864,197 @@ do_distribute() {
 	run_distribute
 }
 
+### namespace: push ###
+
+push_topic() {
+	echo "${content}" >> "${STATE}/topic-names"
+}
+
+push_hashinfo() {
+	echo "${content}" >> "${STATE}/topics"
+}
+
+push_release() {
+	echo "${content}" >> "${STATE}/tagname"
+}
+
+push_topic_range() {
+	[ ! -e "${STATE}/tagname" ] && return
+	[ -e "${STATE}/tagrev-flat" ] && abort "more than one 'topic-range' after 'release'!"
+
+	echo "${content##*..}" > "${STATE}/tagrev-flat"
+}
+
+### command: push ###
+
+resolve_commitish() {
+	${GIT} rev-parse --revs-only "$@" 2>/dev/null
+}
+
+shorten_commitish() {
+	resolve_commitish --short ${1}
+}
+
+resolve_tag() {
+	local remote=$1 tag=$2 commit
+	if [ -n "$remote" ]; then
+		# handles conflicting tags on remote
+		commit=$(git ls-remote -q $remote refs/tags/$tag 2>/dev/null | \
+			sed 's/\s\+.*$//')
+	else
+		commit="refs/tags/$tag"
+	fi
+
+	resolve_commitish "${commit}^{}"
```

Comment:
This is only used to compare the local and remote tags. What's the use-case for accepting different tags with the same commit?